Add Domains V3 resource to CloudFoundryClient

LittleBaiBai opened the pull request:
[#1008]

LittleBaiBai commented:
Hi @twoseat, do you have any feedbacks or concern on this PR?

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java:
```diff
@@ -21,7 +21,7 @@
 import org.cloudfoundry.client.v2.blobstores.Blobstores;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
This is a breaking change - ideally we would have thought ahead and called it `DomainsV2` from the beginning, but we are where we are, so this needs to remain unchanged. Introducing `DomainsV3` and `ReactorDomainsV3` is fine, even if it looks a little inconsistent.

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java:
```diff
@@ -67,7 +68,7 @@
 import org.cloudfoundry.reactor.client.v2.blobstores.ReactorBlobstores;
 import org.cloudfoundry.reactor.client.v2.buildpacks.ReactorBuildpacks;
-import org.cloudfoundry.reactor.client.v2.domains.ReactorDomains;
+import org.cloudfoundry.reactor.client.v2.domains.ReactorDomainsV2;
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java:
```diff
@@ -169,8 +171,14 @@ public DeploymentsV3 deploymentsV3() {
+    @Override
+    @Value.Derived
+    public DomainsV3 domainsV3() {
+        return new ReactorDomainsV3(getConnectionContext(), getRootV3(), getTokenProvider(), getRequestTags());
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java:
```diff
@@ -20,7 +20,7 @@
 import org.cloudfoundry.client.v2.domains.DeleteDomainRequest;
 import org.cloudfoundry.client.v2.domains.DeleteDomainResponse;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java:
```diff
@@ -35,9 +35,9 @@
 
 /**
- * The Reactor-based implementation of {@link Domains}
+ * The Reactor-based implementation of {@link DomainsV2}
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java:
```diff
@@ -35,9 +35,9 @@
+ * The Reactor-based implementation of {@link DomainsV2}
  */
-public final class ReactorDomains extends AbstractClientV2Operations implements Domains {
+public final class ReactorDomainsV2 extends AbstractClientV2Operations implements DomainsV2 {
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java:
```diff
@@ -47,7 +47,7 @@
      * @param requestTags       map with custom http headers which will be added to web request
      */
-    public ReactorDomains(ConnectionContext connectionContext, Mono<String> root, TokenProvider tokenProvider, Map<String, String> requestTags) {
+    public ReactorDomainsV2(ConnectionContext connectionContext, Mono<String> root, TokenProvider tokenProvider, Map<String, String> requestTags) {
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public Mono<ShareDomainResponse> share(ShareDomainRequest request) {
+//        return null;
```
Remove comment

twoseat commented on cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public Mono<Void> unshare(UnshareDomainRequest request) {
+//        return null;
```
Remove comment

twoseat commented on cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/ReactorCloudFoundryClientTest.java:
```diff
@@ -56,8 +56,13 @@ public void builds() {
-    public void domains() {
-        assertThat(this.client.domains()).isNotNull();
+    public void domainsV2() {
+        assertThat(this.client.domainsV2()).isNotNull();
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java:
```diff
@@ -49,9 +49,9 @@
 import static io.netty.handler.codec.http.HttpResponseStatus.OK;
 
-public final class ReactorDomainsTest extends AbstractClientApiTest {
+public final class ReactorDomainsV2Test extends AbstractClientApiTest {
```
Breaking change (see above)

twoseat commented on cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java:
```diff
@@ -49,9 +49,9 @@
+public final class ReactorDomainsV2Test extends AbstractClientApiTest {
 
-    private final ReactorDomains domains = new ReactorDomains(CONNECTION_CONTEXT, this.root, TOKEN_PROVIDER, Collections.emptyMap());
+    private final ReactorDomainsV2 domains = new ReactorDomainsV2(CONNECTION_CONTEXT, this.root, TOKEN_PROVIDER, Collections.emptyMap());
```
Breaking change (see above)

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java:
```diff
@@ -20,7 +20,7 @@
 import org.cloudfoundry.client.v2.blobstores.Blobstores;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Breaking change (see above)

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java:
```diff
@@ -107,9 +108,14 @@
+     * Main entry point to the Cloud Foundry Domains V2 Client API
      */
-    Domains domains();
+    DomainsV2 domainsV2();
```
Breaking change (see above)

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v2/domains/DomainsV2.java:
```diff
@@ -19,9 +19,9 @@
+ * Main entry point to the Cloud Foundry Domains V2 Client API
  */
-public interface Domains {
+public interface DomainsV2 {
```
Breaking change (see above)

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/Domain.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    @JsonProperty("metadata")
+    @Nullable
+    public abstract Metadata getMetadata();
```
Move to alphabetical order

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/DomainsV3.java:
```diff
@@ -0,0 +1,82 @@
+     * @param request the Update a Domain request
+     * @return the response from the Update a Domain request
+     */
+    Mono<UpdateDomainResponse> update(UpdateDomainRequest request);
```
Alphabetical order

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_CreateDomainRequest.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    @JsonProperty("metadata")
+    @Nullable
+    abstract Metadata getMetadata();
```
Alphabetical order

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ListDomainsRequest.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    @FilterParameter("label_selector")
+    @Nullable
+    abstract String getLabelSelector();
```
Alphabetical order

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainRequest.java:
```diff
@@ -0,0 +1,46 @@
+     * The organizations the domain is shared to
+     */
+    @JsonProperty("data")
+    abstract List<Relationship> getData();
```
Alphabetical order

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java:
```diff
@@ -0,0 +1,44 @@
+abstract class _ShareDomainResponse {
+
+    /**
+     * The space
```
Seems this should be `The data`

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Map;
+
+/**
+ * The response payload for the Share Service Instance operation
```
I've made this copy/paste error so many times!

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/OrganizationsV3.java:
```diff
@@ -63,4 +63,12 @@
+     * @param request the List Organization Domains request
+     * @return the response from the List Organization Domains request
+     */
+    Mono<ListOrganizationDomainsResponse> listDomains(ListOrganizationDomainsRequest request); //TODO needs tests
```
Does it still need tests?

LittleBaiBai replied:
Good catch! I added test but forgot to remove the TODO 😅

twoseat commented on cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/_ListOrganizationDomainsRequest.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    @FilterParameter("label_selector")
+    @Nullable
+    abstract String getLabelSelector();
```
Alphabetical order

twoseat commented on cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java:
```diff
@@ -19,7 +19,7 @@
 import org.cloudfoundry.client.v2.applications.ApplicationsV2;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Breaking change (see above)

twoseat commented on cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java:
```diff
@@ -86,7 +86,7 @@
     protected final CloudFoundryClient cloudFoundryClient = mock(CloudFoundryClient.class, RETURNS_SMART_NULLS);
 
-    protected final Domains domains = mock(Domains.class, RETURNS_SMART_NULLS);
+    protected final DomainsV2 domainsV2 = mock(DomainsV2.class, RETURNS_SMART_NULLS);
```
Breaking change (see above)

twoseat commented on cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java:
```diff
@@ -149,7 +149,7 @@ public final void mockClient() {
         when(this.cloudFoundryClient.applicationsV3()).thenReturn(this.applicationsV3);
         when(this.cloudFoundryClient.buildpacks()).thenReturn(this.buildpacks);
-        when(this.cloudFoundryClient.domains()).thenReturn(this.domains);
+        when(this.cloudFoundryClient.domainsV2()).thenReturn(this.domainsV2);
```
Breaking change (see above)

LittleBaiBai commented:
@twoseat Thanks for your thorough review! I believe I have addressed all the comments you left and made sure no `DomainsV2` showed up in a global search. 

I also added integration tests, but I may need some help with getting one of them to pass - `DomainsTest.delete`. This test fails because `delete` is an asynchronous operation and the API spec declared the response to have job location url, but the CF instance I'm using is not returning it. So I wasn't able to wait until job is done as in some other integration test, then the assertion right after deletion fails. We could add retries or arbitrary wait time, but I'd love to hear your thoughts on this first.

twoseat commented:
Thanks for the updates @LittleBaiBai - I'm working on it now, and have found what I guess is your domains issue (there's an undocumented `warnings` field returned as part of the `job`, and we're configured to choke on unexpected fields for test purposes). I've added something temporarily that makes the test work, but this issue (and possibly other V3 stuff) might have to wait until I get some feedback on what the field is supposed to contain.
